source/language/openpulse.rst

```diff
@@ -0,0 +1,317 @@
+OpenPulse Grammar
+=================
+
+In addition to OpenQASM instructions, ``defcal`` blocks may contain OpenPulse 
+instructions. Certain OpenPulse instructions are further restricted to only
+appearing inside of a ``defcal`` block.
+
+These instructions are motivated by the original OpenPulse specification which
+was defined as a JSON wire-format for pulse-level quantum programs in the paper
+`Qiskit Backend Specifications for OpenQASM and OpenPulse Experiment <https://arxiv.org/abs/1809.03452>`_. 
+The text format described here has several advantages over the equivalent JSON 
+format:
+
+- It is more readable
+- Absolute time is handled through built-in timekeeping instructions
+- Gates and classical instructions can be mixed in with the pulses to create far richer calibrations
+- Pulse definitions are tied to circuit instructions rather than circuit programs
+- Richer handling of frames
+
+Channels
+--------
+
+Channels map to a hardware resource, which can play pulses to manipulate a qubit
+or capture signal from the qubit to perform a measurement. There is a many-to-many
+relationship between qubits and channels. One qubit may have multiple channels
+connecting to it. Pulses on different channels would have different physical
+interactions with that qubit. A channel may also have many qubits. For instance,
+a channel could manipulate the coupling between two neighboring qubits, or
+could even reference multiple qubits in a chain.
+
+Channels are vendor-specific and device-specific. It is expected that vendors
+of quantum hardware provide the appropriate channel names and qubit mappings
+as configuration information to end users. Each channel may also have associated 
+static settings, such as local-oscillator frequencies, which do not vary 
+throughout program execution. Again it is expected that vendors of quantum 
+hardware provide a method for manipulating those static settings if appropriate.
+
+There are two kinds of channels: transmit channels (sending input to a quantum
+device) and receive channels (reading output from a quantum device).
+
+Interacting with channels directly is onerous and makes the pulse programs less
+portable. The OpenPulse defcalgrammar does not directly work with channels and
+instead works with abstract "frames", described below. The static settings above 
+should also include a mapping of these abstract frames to the correct channels.
```

Comment:
One problem I see with restricting events to frames is an assumption that you only wish to schedule events (e.g. play or capture) on something that tracks phase. There are instances in which I need deterministic scheduling outside this paradigm. Some examples include wanting to schedule a capture at a very specific time that maps to a camera or a single-photon detector acquisition, or schedule a play that tunes a magnetic field coil during a gate. Although one could argue that these would just have a frequency and phase of 0, it seems funny to have to over-specify such properties -- I would probably ask why I can't just use a channel.